### DISCUSSION TITLE
想问问如何做全局的登录操作

### DISCUSSION THREAD

**Discussion by cdoky** [yangKJ/RxNetworks#4] [General]
#### 场景
某个页面有5个接口，其中3个接口是有登录验证的，
- 如果未登录的话，后台会返回401，然后需要弹出登录页面进行登录操作
- 登录完成后，3个接口需要携带token重新请求
- 登录失败，则提示错误

#### 我的解决方案
我没用过Moya，之前是直接用Alamofire，然后通过拦截器去实现的

想问问，Moya这一块，有没有更好的方法去实现？
还是说也是一样需要通过Alamofire的拦截器去实现

**Comment by yangKJ** [comment]
我会这样设计一款插件，
1.在最开始请求配置时刻`prepare`注入token
2.在最后返回时刻去拦截是否为401，
如果为401则用信号量去卡住这次网络然后去处理登录相关逻辑拿到新的token之后解开信号量，然后自动重新请求该网络

**Reply by cdoky** [reply]
<img width="1009" alt="image" src="https://user-images.githubusercontent.com/626401/153699734-58fd78a6-f84d-4af6-afc0-b1e21825acb7.png">

`public func lastNever(_ tuple: LastNeverTuple, target: TargetType) -> LastNeverTuple`

这个方法是在主线程执行的，
在这里使用信号量会卡住主线程
登陆操作无法进行

当前只是测试阶段，想要所有的请求，都触发登陆操作
stateCode逻辑可以暂时不管

**Comment by yangKJ** [comment]
不好意思，才注意到你这边的回复。
这个问题其实很简单，你这边是不是在使用的时候未使用`callbackQueue`这个参数，
这边我其实还是只是对moya的二次封装，moya的`request`方法里面的`callbackQueue = nil`默认`DispatchQueue.main`
所以你这边其实可以把`callbackQueue`这个参数使用起来即可解决你的问题。
或者我回头看看有没有必要单开一条子线程去处理这个结果- -！

**Reply by cdoky** [reply]
对，我现在把plugin的那几个操作都放子线程运行就可以了

**Comment by yangKJ** [comment]
嗯嗯，所以你的问题得到解决了嘛？
我刚刚重新整理了一下，我这边会默认注入一个子线程进去。
这样就如果`callbackQueue`参数不使用，也就默认使用子线程。

**Reply by cdoky** [reply]
嗯，已经解决了。非常感谢
